### PR TITLE
Give Site Administrator permission to manage users

### PR DESCRIPTION
**Pull request by wesleybl**
* it was necessary to give permission `plone.restapi: Access Plone user information` and `Manage users` to Site Administrator.
* An upgrade step has not yet been carried out to give these permissions to an already created portal. I don't know if I should do this here in `plone.restapi` or in `plone.app.upgrade`.
* Several checks were made to prevent the Site Administrator from giving Manager permission to any user.
* I will do a PR on Volto, to remove options from the Site Administrator to give Manager permission. To do this, roles, groups and users endpoints send information relevant to this.
* A good way to test blocking the Site Administrator from giving Manager permission is to use this PR on the current Volto. The options are still available in the Volto, but the backend blocks them.
* Maybe reviewing the commits might be easier.

Fixes plone/volto#3917

### CODE REVIEW

**Comment by wesleybl** [comment]
@jenkins-plone-org please run jobs

**Comment by erral** [comment]
how does this work in classic?

**Comment by wesleybl** [comment]
> how does this work in classic?

In Plone Classic, users with the Site Administrator role are allowed to manage users. See:

https://github.com/plone/Products.CMFPlone/blob/1c1a0445a4a042af20a3602dda629999a8bf1186/Products/CMFPlone/profiles/default/rolemap.xml#L194-L199

**Review by jensens — APPROVED** [review]

**Comment by jensens** [comment]
I would appreciate a second review before merge, since this is a security issue.

**Comment by davisagli** [comment]
I will review this, but I want to take the time to do it carefully, so I haven't gotten to it yet.

**Comment by wesleybl** [comment]
> I will review this, but I want to take the time to do it carefully, so I haven't gotten to it yet.

@davisagli could you please take a look here? I really need this. On our Sites we do not provide Manager permission to clients, but they need to manage users.

File: src/plone/restapi/profiles/default/rolemap.xml
Comment (davisagli) [inline src/plone/restapi/profiles/default/rolemap.xml:16]: @wesleybl I see that in the rest of the PR, you've mostly changed the endpoints to check plone.app.controlpanel.UsersAndGroups rather than "Manage users". Is there something that "Manage users" is still needed for?

If so we'll need to take a closer look at how to solve it, because I don't think we can safely give the "Manage users" permission to Site Administrators. If I'm not mistaken, it provides access to some methods outside of our control (i.e. in PAS and the base Zope user folder) which could be used to escalate privileges by adding the Manager role to oneself.

If you can point me to where it's still preventing a Site Admin from doing something with the rest of your changes in place, I can try to give advice on how to fix it.

(Side note: if adding this permission were the correct thing to do, it should happen in CMFPlone rather than here. It's too confusing if a role has different permissions depending on whether or not plone.restapi is installed.)

File: src/plone/restapi/profiles/default/rolemap.xml
Comment (davisagli) [inline src/plone/restapi/profiles/default/rolemap.xml:12]: This one is okay, since the permission is defined here in plone.restapi. But we do need an upgrade step. I think it's fine for the upgrade step to also go in plone.restapi.

**Review by davisagli — CHANGES_REQUESTED** [review]
@wesleybl First, apologies for my delay in reviewing this.

Second, great work on a complicated set of use cases. This will also be useful in some of the sites I work on.

I'm adding a bunch of comments; most of them are pretty minor. There is one important question about why the "Manage users" permission is still needed, because that's not immediately clear to me (maybe it would be if I actually started an instance and tried it!) and I think we need to avoid it.

File: src/plone/restapi/profiles/default/rolemap.xml
Comment (davisagli) [reply src/plone/restapi/profiles/default/rolemap.xml:16]: @wesleybl Ah, I found your comment about this here: https://github.com/plone/volto/issues/3917#issuecomment-1721836345

At first glance, I think we should resolve that by making plone.restapi also use the deleteMembers function from CMFPlone which avoids the "Manage users" check in CMFCore. The function even has a comment to remind us that that's what it is there for: https://github.com/plone/Products.CMFPlone/blob/bad7ee61597c72f165f5db979ee04efd06066d22/Products/CMFPlone/controlpanel/browser/usergroups_usersoverview.py#L225-L227

File: src/plone/restapi/services/users/get.py
Comment (wesleybl) [reply src/plone/restapi/services/users/get.py:200]: Today a Manager can delete himself. Are we going to change this? If so, this can be done in another PR right?

Comment (davisagli) [reply src/plone/restapi/services/users/get.py:200]: @wesleybl Yeah, we can leave this out of the current PR. I can imagine reasonable arguments for either yes or no.

File: src/plone/restapi/services/roles/get.py
Comment (davisagli) [reply src/plone/restapi/services/roles/get.py:41]: @wesleybl It doesn't make sense to me either, and it feels like an inconsistency that is more likely there by accident than by design. I think we can decide to do something different in the REST API for the sake of usability, and always allow assigning any role (except Manager when being assigned by non-Managers). I would prefer to do that than add both `can_assign` and `can_assign_add`, since it's hard to explain and locks us into supporting that in the API indefinitely.

File: src/plone/restapi/profiles/default/rolemap.xml
Comment (wesleybl) [reply src/plone/restapi/profiles/default/rolemap.xml:12]: My question is whether this upgrade will not be lost in plone.restapi. Generally when we update Plone, we do not check the upgrade step on a product-by-product. If we run the Plone upgrade, will this upgrade here be run?

Comment (davisagli) [reply src/plone/restapi/profiles/default/rolemap.xml:12]: @wesleybl The perhaps surprising answer is: yes. plone.app.upgrade has a list of specific packages (core addons) whose upgrade steps will also be run when using the @@plone-upgrade view, and it includes plone.restapi: https://github.com/plone/Products.CMFPlone/pull/3457

Comment (wesleybl) [reply src/plone/restapi/profiles/default/rolemap.xml:12]: Cool

File: src/plone/restapi/profiles/default/rolemap.xml
Comment (wesleybl) [reply src/plone/restapi/profiles/default/rolemap.xml:16]: @davisagli the delete users service allows you not to delete [memberareas](https://github.com/plone/plone.restapi/blob/48d2569b00b4e166556b6699ce4ee086f30a2c7e/src/plone/restapi/services/users/delete.py#L34-L36) and [localroles](https://github.com/plone/plone.restapi/blob/48d2569b00b4e166556b6699ce4ee086f30a2c7e/src/plone/restapi/services/users/delete.py#L38-L40). `deleteMembers` [always does this](https://github.com/plone/Products.CMFPlone/blob/bad7ee61597c72f165f5db979ee04efd06066d22/Products/CMFPlone/controlpanel/browser/usergroups_usersoverview.py#L252C33-L261). Should we keep this option in the service? If so, you will not be able to use `deleteMembers` directly.

Comment (davisagli) [reply src/plone/restapi/profiles/default/rolemap.xml:16]: @wesleybl We should keep those options. We can copy a portion of the code from `deleteMembers` into the plone.restapi user delete service. It's a small amount of code duplication but probably worth it to avoid needing to deal with the complexity of making changes to `deleteMembers` in multiple branches of CMFPlone.

Comment (wesleybl) [reply src/plone/restapi/profiles/default/rolemap.xml]: @davisagli I kept it. Please take a look at https://github.com/plone/plone.restapi/pull/1712/commits/51eab0caef415490b46839f45b97fa680e42f972

File: src/plone/restapi/services/roles/get.py
Comment (wesleybl) [reply src/plone/restapi/services/roles/get.py:41]: @davisagli It won't be easy to do this. See my comment at: https://github.com/plone/volto/issues/3917#issuecomment-1730277651

File: src/plone/restapi/services/users/update.py
Comment (wesleybl) [reply src/plone/restapi/services/users/update.py]: Done.

File: src/plone/restapi/services/groups/add.py
Comment (wesleybl) [reply src/plone/restapi/services/groups/add.py]: Done.

File: src/plone/restapi/services/groups/delete.py
Comment (wesleybl) [reply src/plone/restapi/services/groups/delete.py]: Done.

File: src/plone/restapi/services/groups/update.py
Comment (wesleybl) [reply src/plone/restapi/services/groups/update.py]: Done.

File: src/plone/restapi/services/users/delete.py
Comment (wesleybl) [reply src/plone/restapi/services/users/delete.py]: Done.

File: src/plone/restapi/services/roles/get.py
Comment (davisagli) [reply src/plone/restapi/services/roles/get.py:41]: @wesleybl Okay, I think I understand: CMFCore's RegistrationTool.addMember enforces this, and the plone.restapi user add service currently uses it.

It doesn't look terribly hard to work around to me: we could make the user add service first add a member without specifying additional roles, and then separately add the roles the same way that it is done when updating an existing user.

I'm okay with calling this out of scope for the current PR, but I still don't really want to add `can_assign_add` to the API, since in the long run I think we should fix it so that there's no difference in which roles can be specified when adding vs editing a user.

File: src/plone/restapi/services/groups/add.py
Comment (wesleybl) [reply src/plone/restapi/services/groups/add.py:33]: I do

```python
roles = data.get("roles", [])
```

If it's not a list, it would give an error anyway.

File: news/1712.feature
Comment (wesleybl) [reply news/1712.feature]: Done.

File: src/plone/restapi/services/roles/get.py
Comment (wesleybl) [reply src/plone/restapi/services/roles/get.py]: @davisagli I removed the key `can_assign_add`

File: src/plone/restapi/profiles/default/rolemap.xml
Comment (wesleybl) [reply src/plone/restapi/profiles/default/rolemap.xml:12]: I created the upgrade step.

File: src/plone/restapi/services/groups/get.py
Comment (wesleybl) [reply src/plone/restapi/services/groups/get.py]: Done.

**Comment by wesleybl** [comment]
@davisagli I made all the requested changes. Can you take a look please?

**Comment by davisagli** [comment]
@wesleybl @gforcada I think it's intentional that the Jenkins build is not triggered for 3.11 -- we run on the oldest and latest supported Python versions only, to manage the load. I just need to update the repository branch protection settings so that the 3.12 check is required instead of the 3.11 check, and then the 3.11 check will not show up here.

**Review by davisagli — CHANGES_REQUESTED** [review]
@wesleybl This looks really close! I just noticed that we should probably only add `can_delete` in the serializers if the current user has permission to do those actions.

File: src/plone/restapi/serializer/user.py
Comment (wesleybl) [reply src/plone/restapi/serializer/user.py:48]: @davisagli analyzing your comment, I noticed that the user service uses the permissions [Plone Site Setup: Users and Groups](https://github.com/plone/plone.restapi/blob/db7d5dffe18d39d7487b484f5124b0abf48cfa9e/src/plone/restapi/services/users/get.py#L178-L184), which is from Plone, and [plone.restapi: Access Plone user information](https://github.com/plone/plone.restapi/blob/db7d5dffe18d39d7487b484f5124b0abf48cfa9e/src/plone/restapi/services/users/get.py#L186-L190), which is from `plone.restapi`. I wonder if we couldn't just use Plone's. They have the same roles now.

Comment (wesleybl) [reply src/plone/restapi/serializer/user.py:48]: This was @sneridagh' PR who created the permission: https://github.com/plone/plone.restapi/pull/525

**Comment by wesleybl** [comment]
@jenkins-plone-org please run jobs

File: src/plone/restapi/serializer/user.py
Comment (wesleybl) [reply src/plone/restapi/serializer/user.py]: >  So I think we should only add the can_delete key if the current user has the plone.app.controlpanel.UsersAndGroups permission.

Done

File: src/plone/restapi/serializer/group.py
Comment (wesleybl) [reply src/plone/restapi/serializer/group.py]: Done

File: src/plone/restapi/serializer/user.py
Comment (davisagli) [reply src/plone/restapi/serializer/user.py]: > I noticed that the user service uses the permissions [Plone Site Setup: Users and Groups](https://github.com/plone/plone.restapi/blob/db7d5dffe18d39d7487b484f5124b0abf48cfa9e/src/plone/restapi/services/users/get.py#L178-L184), which is from Plone, and [plone.restapi: Access Plone user information](https://github.com/plone/plone.restapi/blob/db7d5dffe18d39d7487b484f5124b0abf48cfa9e/src/plone/restapi/services/users/get.py#L186-L190), which is from plone.restapi. I wonder if we couldn't just use Plone's. They have the same roles now.

@wesleybl It could be a breaking change if someone has customized the roles for that permission. It may be worth more thought but I would leave it out of this PR.

(I don't consider the change from `Manage portal` to `Plone Site Setup: Users and Groups` to be breaking, because `Manage portal` should only ever be assigned to Managers, who should also always have `Plone Site Setup: Users and Groups`.)

**Comment by wesleybl** [comment]
@jenkins-plone-org please run jobs

File: src/plone/restapi/permissions.py
Comment (wesleybl) [reply src/plone/restapi/permissions.py]: Done.

**Review by davisagli — APPROVED** [review]

File: src/plone/restapi/tests/test_services_roles.py
Comment (ksuess) [inline src/plone/restapi/tests/test_services_roles.py:100]: You may want to use 
setRoles(self.portal, TEST_USER_ID, ["Site Administrator"])

**Review by ksuess — CHANGES_REQUESTED** [review]
The _Site Administrator_ role should have the permission to access users information, OK, so "plone.restapi: Access Plone user information" is now for Manager and _Site Administrator_. This is a needed change.
But the rest of the PR extends the @roles endpoint with permission info. Which I think is the door to dilute the purity of the concept of the endpoints.

**can_assign**
The purpose of can_assign (which is BTW a role_can_be_assigned_by_current_user) seems to be to distingusih between Manager and Site Administrator, to avoid that a Site Administrator can manage (and delete) Managers.
I think this is not the job of the service _@roles_
IMHO, the user control panel and groups control panel should know which role the current user has and act accordingly.
So the question seems to be, how the frontend can know the role of the current user.

File: src/plone/restapi/serializer/user.py
Comment (wesleybl) [reply src/plone/restapi/serializer/user.py:37]: @ksuess today the `can_delete` key is only returned if the current user can manage user. See:

https://github.com/plone/plone.restapi/blob/08f8f871bbf63c4f1033cc3f2c46d371e994aa22/src/plone/restapi/serializer/user.py#L54-L55

I can test `self.can_manage_users` inside the `can_delete` method too, but it would be strange because I would have to test it twice. What is your opinion?

File: src/plone/restapi/tests/test_services_roles.py
Comment (wesleybl) [reply src/plone/restapi/tests/test_services_roles.py:100]: @ksuess `self.api_session.get("/@roles")`. `self.api_session` already has the Manger authentication header:

https://github.com/plone/plone.restapi/blob/08f8f871bbf63c4f1033cc3f2c46d371e994aa22/src/plone/restapi/tests/test_services_roles.py#L21

File: src/plone/restapi/permissions.py
Comment (wesleybl) [reply src/plone/restapi/permissions.py:6]: Done.

File: src/plone/restapi/services/users/add.py
Comment (wesleybl) [reply src/plone/restapi/services/users/add.py]: Done.

**Comment by wesleybl** [comment]
> IMHO, the user control panel and groups control panel should know which role the current user has and act accordingly.
So the question seems to be, how the frontend can know the role of the current user.

@ksuess yes. The frontend does not know what the current user's roles are. So I thought it would be better to do it this way. Do you have any suggestions for doing this differently?

We could have a "@current_user_roles" endpoint, but I don't like that idea. It would be more complex and less informative.

@davisagli  opinions?

**Comment by tisto** [comment]
FYI: @davisagli is on holiday until mid-next week...

**Comment by davisagli** [comment]
@wesleybl I think @ksuess has a reasonable point that "can_assign" in the `@roles` endpoint is really more information about the current user's abilities than about the role.

We should already have the current user's roles in the frontend redux state -- the PersonalTools component fetches the user data from /@users/[userid] using the userid decoded from the auth JWT token, and the response includes the user's global roles. https://6.docs.plone.org/plone.restapi/docs/source/endpoints/users.html#read-user

File: src/plone/restapi/serializer/user.py
Comment (ksuess) [reply src/plone/restapi/serializer/user.py:37]: Enhancing data with an additional key value pair for all cases is more reliable and better to handle by the consumer.

```
class BaseSerializer:
    def __init__(self, context, request):
        self.context = context
        self.request = request

    def _can_be_managed_by_current_user(self, roles):
        if check_permission(ManagePortal, self.context):
            return True
        # Allow Site Administrator to manage users that don't have Manager role.
        return (
            check_permission(PloneManageUsers, self.context) and "Manager" not in roles
        )

    def __call__(self):
        user = self.context
        portal = getSite()

        # Roles of context (to be serialized user)
        roles = user.getRoles()
        # Anonymous and Authenticated are pseudo roles assign automatically
        # to logged-in or logged-out users. They should not be exposed here.
        roles = sorted(list(set(roles) - {"Anonymous", "Authenticated"}))

        data = {
            "@id": f"{portal.absolute_url()}/@users/{user.id}",
            "id": user.id,
            "username": user.getUserName(),
            "roles": roles,
            "can_be_managed_by_current_user": self._can_be_managed_by_current_user(
                roles
            ),
        }
```

**Comment by ksuess** [comment]
> @wesleybl I think @ksuess has a reasonable point that "can_assign" in the `@roles` endpoint is really more information about the current user's abilities than about the role.
> 
> We should already have the current user's roles in the frontend redux state -- the PersonalTools component fetches the user data from /@users/[userid] using the userid decoded from the auth JWT token, and the response includes the user's global roles. https://6.docs.plone.org/plone.restapi/docs/source/endpoints/users.html#read-user

current user's roles are not fetched unless the personal tools are visited. But the fetch of the user data of the authenticated user can be done in the users control panel component. This would make the change of _@roles_ endpoint dispensable. The users control panel component can act according the roles of the authenticated user: 
- disable checkboxes and hide delete menu for users with Manager role if the authenticated user is only Site Administrator.
- disable checkboxes for Manager role if the authenticated user is only Site Administrator.
grafik

File: src/plone/restapi/tests/test_services_roles.py
Comment (ksuess) [reply src/plone/restapi/tests/test_services_roles.py:100]: OK. I don't have a strong opinion here.

**Comment by wesleybl** [comment]
@ksuess @davisagli Okay. Let's continue like this then. Although I think we would be transferring responsibility from the backend to the frontend.

If we have the authenticated user's roles on the frontend, we can remove the `can_assign` and `can_delete` keys from the endpoints.

File: src/plone/restapi/serializer/user.py
Comment (wesleybl) [reply src/plone/restapi/serializer/user.py:37]: If we have the authenticated user roles in the frontend, we can remove the can_assign and can_delete keys from the endpoints. I will remove them.

Comment (wesleybl) [reply src/plone/restapi/serializer/user.py]: @ksuess I removed the `can_delete key` from the users endpoint.

**Comment by wesleybl** [comment]
@ksuess @davisagli I removed the `can_delete` and `can_assign` key from endpoints. Can you check if it's ok now?

**Comment by wesleybl** [comment]
@jenkins-plone-org please run jobs

**Comment by davisagli** [comment]
@wesleybl Thanks -- this is on my list to review, but I'm also currently on a holiday break. Please remind me about it in a week if I haven't gotten to it yet.

File: src/plone/restapi/serializer/utils.py
Comment (davisagli) [inline src/plone/restapi/serializer/utils.py:59]: Looks like this is no longer used anywhere

**Review by davisagli — APPROVED** [review]
This looks good to me now!

**Comment by davisagli** [comment]
@jenkins-plone-org please run jobs

**Comment by davisagli** [comment]
@wesleybl Thanks for your hard work and patience on this one!

**Comment by tisto** [comment]
@wesleybl thank you for your hard work on this!

@davisagli just to double-check: this is a non-breaking feature, correct? I can cut a new release...

**Comment by davisagli** [comment]
@tisto yes, as far as I can tell. It adds the ability for site admins to do some user management actions that were previously only available for Managers

**Comment by tisto** [comment]
@wesleybl @davisagli released https://pypi.org/project/plone.restapi/9.3.0/

**Pull request merged**
